packages/theming-core/src/generics/themeRegistry.ts

```diff
@@ -0,0 +1,247 @@
+   * not specified the default theme will be returned
+   */
+  getTheme: IGetTheme<ITheme>;
+
+  /**
+   * update the platform defaults.  This is less likely to happen on web but on native
+   * the system settings might change requiring all dependent themes to be updated.
+   */
+  updatePlatformDefaults: IUpdatePlatformDefaults<ITheme>;
+
+  /**
+   * The actual graph of named themes
+   */
+  graph: object;
+}
+
+/**
+ * Creates a theme registry object which caches theme information and provides typed functions
+ * for working with themes in the registry.
+ * @param platformDefaults - A fully defined theme with default values for the given platform
+ * @param resolver - A function which takes a partial theme, a fully defined parent theme, and
+ * produces a fully resolved theme.
+ *
+ * @internal This is an experimental interface and will be changed post design review.
+ */
+export function createThemeRegistry<ITheme, IThemeDefinition>(
+  platformDefaults: ITheme,
+  resolver: ResolveTheme<ITheme, IThemeDefinition>
+): IThemeRegistry<ITheme, IThemeDefinition> {
+  const graph: IThemeGraph<ITheme, IThemeDefinition> = {
+    _platform: { resolved: platformDefaults },
+    default: { parent: _platformKey }
+  };
+  return {
+    graph,
+    registerTheme: (definition: IThemeDefinition | ProcessTheme<ITheme, IThemeDefinition>, name?: string, parent?: string): void => {
+      _registerTheme<ITheme, IThemeDefinition>(graph, definition, name, parent);
+    },
+    getTheme: (name?: string): ITheme => {
+      return _getTheme<ITheme, IThemeDefinition>(graph, resolver, name);
+    },
+    updatePlatformDefaults: (newDefaults: ITheme): void => {
+      _updatePlatformDefaults<ITheme, IThemeDefinition>(graph, newDefaults);
+    }
+  };
+}
+
+/**
+ * Register a new theme by name.  This will invalidate the portion of the graph that is dependent on the new
+ * theme.  This will error on bad dependencies including cycles.
+ *
+ * @param graph - theme graph used for tracking dependencies
+ * @param definition - new theme definition or processing function to set in the theme
+ * @param name - name of the theme, if empty will be treated as 'default'
+ * @param parent - optional theme parent, if empty will be treated as 'default' if not default theme
+ */
+function _registerTheme<ITheme, IThemeDefinition>(
+  graph: IThemeGraph<ITheme, IThemeDefinition>,
+  definition: IThemeDefinition | ProcessTheme<ITheme, IThemeDefinition>,
+  name?: string,
+  parent?: string
+): void {
+  name = name || _defaultKey;
+  parent = (name === _defaultKey && _platformKey) || parent || _defaultKey;
+  if (_wouldCauseCycle(graph, name, parent)) {
+    throw new Error('Attempt to register a dependent theme that would cause a cycle');
+  } else if (!graph[parent]) {
+    throw new Error('Attempting to parent to an unknown theme');
+  } else if (name === _platformKey) {
+    throw new Error('A platform theme cannot be registered, call updatePlatformDefaults instead');
```

Comment:
Would we really want to throw an exception for all of these cases? In general we seem to very rarely throw exceptions in Fabric, but I also haven't really seen a prescribed approach for throwing exceptions. In this case, I wonder if it would be better to log a warning but no-op for some of these rather than blow up the application.